server/routes/calendar.js

```diff
@@ -112,4 +113,65 @@ router.get("/availability/:userId", (req, res) => {
   );
 });
 
+const coverColors = ["#ffff42", "#ffaf0f", "#ffc34d", "#4eb548"];
+
+router.post("/openshifts/:userId", (req, res) => {
+  let shifts = req.body.items;
+  pool.query(
+    "select * from coverrequests full join shifts on coverrequests.shift_id = shifts.shift_id where coverer_id is not distinct from $1 and request_id is distinct from $1",
+    [null],
+    (error, result) => {
+      if (error) {
+        console.log(error);
+        throw error;
+      }
+      let shiftid_to_color = {};
+      for (var j = 0; j < result.rows.length; j += 1) {
+        let currentRow1 = result.rows[j];
+        if (!(currentRow1.shift_id in shiftid_to_color)) {
+          shiftid_to_color[currentRow1.shift_id] = coverColors[j % 4];
+        }
+        for (var i = 0; i < 168; i += 1) {
+          let sameStartEndValid =
+            shifts[i].day == currentRow1.start_time.getDay() &&
+            shifts[i].start >= currentRow1.start_time.getHours() &&
+            shifts[i].end <= currentRow1.end_time.getHours();
+          let diffStartEndValid =
+            currentRow1.start_time.getDay() != currentRow1.end_time.getDay() &&
+            ((shifts[i].day == currentRow1.start_time.getDay() &&
+              shifts[i].start >= currentRow1.start_time.getHours()) ||
+              (shifts[i].day == currentRow1.end_time.getDay() &&
+                shifts[i].end <= currentRow1.end_time.getHours()));
+          if (
+            (sameStartEndValid || diffStartEndValid) &&
+            currentRow1.sle_id != req.body.userId
```

Comment:
For some reason this isn't ignoring coverrequests from the user that is viewing it. Let's just move that logic into the sql query.